lib/cn/cn_tree.c

```diff
@@ -920,21 +920,28 @@ cn_node_comp_token_put(struct cn_tree_node *tn)
 static void
 cn_comp_release(struct cn_compaction_work *w)
 {
-    bool join = (w->cw_action == CN_ACTION_JOIN);
+    const bool join = (w->cw_action == CN_ACTION_JOIN);
     assert(w->cw_node);
 
     if (w->cw_err || join) {
         struct kvset_list_entry *le;
-        uint kx, cnt = join ? cn_ns_kvsets(&w->cw_node->tn_ns) : w->cw_kvset_cnt;
+        uint cnt = join ? cn_ns_kvsets(&w->cw_node->tn_ns) : w->cw_kvset_cnt;
 
         /* unmark input kvsets */
         le = join ? list_last_entry(&w->cw_node->tn_kvset_list, typeof(*le), le_link) : w->cw_mark;
-        for (kx = 0; kx < cnt; kx++) {
+        for (uint kx = 0; kx < cnt; kx++) {
```

Comment:
This code entire function looks pretty sketch to me, it could be messing around with a live node anytime err is not zero.
Consider, the action fails and then this thread gets put to sleep after releasing the compaction token on the node(s).  The say a join job run which is successful and finishes deleting the node.  When the original thread wakes up the node is gone.
I see that cn_comp_cleanup() seems to be careful not to touch the node, except for printing the node ID if there was an error (so, same problem there, it shouldn't do that).
We can probably merge cn_comp_release() into cn_comp_cleanup() and deal with it there, by not releasing the compaction token until you're actually done messing with the node..

Reply:
Yeah, I think the cleanest solution here is to move the busycnt and token changes in cn_comp_cleanup() to the end of the function after error handling, and then merge cn_comp_release() into cn_comp_cleanup().
For all actions other than split/join the node could receive spills, so you'll need the read tree lock to unmark kvsets to keep the kvset list stable while you're walking it...

Reply:
That's a great catch, Greg! I incorrectly assumed that it is safe to unmark the kvsets here for join as we already do that today in the error case for other compactions, my bad.

Will address this in my next push.

Reply:
Greg, I have moved the busycnt and dropping token to the end of cn_comp_cleanup().
Also, I'm now calling cn_comp_release() before cn_comp_cleanup(), so the kvsets can be unmarked safely...